### PR TITLE
Overhaul of trace ID identification and sending

### PR DESCRIPTION
**Pull request by kitsuneninetails**
Trace ID cannot be identified per thread, because it is possible for a
thread to service many traces, if there is any work-stealing amongst the
threads (if one "task" running a trace awaits, the thread could
suspend that task and grab another, starting a new trace).

Since it is impossible for the datadog tracer to know the trace ID in
this circumstance, we must force the user to tell us the ID through the
span attributes (key/value pairs in the span! macro call).  Likewise for
the event to send to datadog, as this even can occur outside of any of
the spans for the trace the user wishes to send (i.e. once the trace is
complete, the user asks for the tradce to be shipped off to datadog;
this is a common use case).

Once the user qualifies the trace ID for a new span, we can internally
keep track of which trace each thread is currently servicing based off
of the "enter/exit" span calls (one thread can only "ENTER" a span one
at a time; they cannot ever be "in" two spans at once).  Keeping track
of which threads are in which traces (and vice versa), any span
operations and/or log operations can look up the "current" span based on
the thread ID alone.

Also, a fix was made for the timing of the span durations, as they were 
being decided by the event loop.  This made some spans run a little longer
than expected (because of the time between the true span end and when the
event loop finally got around to processing the event), and disrupted the
flame graphs, gave wrong durations, etc.  Instead, send the end time as a
part of the event, set by the thread that actually ends the span, so that it is
more accurate.

Furthmore, add support for APM analytics by adding an "APM Config"
option to the main config.  Turning on the apm_config (using apm_enabled
flag) will send APM Analytics to datadog, allowing APM metrics to be
tallied and used (these get set in the tags of each span).

### CODE REVIEW

File: tmp [hunk 1]
```diff
@@ -0,0 +1,3255 @@
+/home/micucci/.cargo/bin/cargo test --color=always --package api-consumers --test load_test test_blast -- --nocapture --exact
```
Comment (craigpastro) [inline tmp]: Should this file be included?

File: Cargo.toml [hunk 2]
```diff
@@ -21,4 +21,4 @@ tracing-futures = "*"
 
 [dev-dependencies]
 rand = "0.3"
-tokio = { version = "0.2", features = ["rt-core", "sync", "rt-threaded", "macros"] }
\ No newline at end of file
+tokio = { version = "0.2", features = ["full"] }
```
Comment (craigpastro) [inline Cargo.toml]: Missing newline

**Review by craigpastro — DISMISSED** [review]
Nice! LGTM 👍 I think your `tmp` file snuck in.

File: tmp (hunk 1 above)
Comment (kitsuneninetails) [reply tmp]: Nope! :)

**Review by craigpastro — APPROVED** [review]

**Pull request merged**